m2cgen/interpreters/mixins.py

```diff
@@ -36,10 +36,10 @@ def _pre_interpret_hook(self, expr, bin_depth=0, **kwargs):
 
     # Default implementation. Simply adds new variable.
     def bin_depth_threshold_hook(self, expr, **kwargs):
-        var_name = self._cg.add_var_declaration(expr.output_size)
+        if expr in self._cached_expr_results:
+            return self._cached_expr_results[expr].var_name
         result = self._do_interpret(expr, **kwargs)
-        self._cg.add_var_assignment(var_name, result, expr.output_size)
-        return var_name
+        return self._cache_reused_expr(expr, result)
```

Comment:
I don't think we should mix 2 cases together. We have 2 scenarios to consider here:
1. An expression that is explicitly marked as reusable (`to_reuse` is true). This is what we call "caching" in the code.
2. An AST requires splitting due to limitation of the target platform / langauge.

I can easily see how a specific interpreter may choose to opt into one of these mechanisms but not the other. I can also see how 2 of these may require different implementations (eg. new variable vs new function). Additionally by default the caching is disabled (https://github.com/BayesWitnesses/m2cgen/blob/892d350d114ac81696522c41c1498ebc2c8f1b28/m2cgen/interpreters/interpreter.py#L49-L51), which may be quite unexpected if one chooses to rely on `BinExpressionDepthTrackingMixin`.

Reply:
Absolutely agree with you! Refactored.